onnxruntime/core/providers/dml/dml_provider_factory.cc

```diff
@@ -92,10 +95,286 @@ bool IsSoftwareAdapter(IDXGIAdapter1* adapter) {
     return isSoftwareAdapter || (isBasicRenderDriverVendorId && isBasicRenderDriverDeviceId);
 }
 
+static bool IsHardwareAdapter(IDXCoreAdapter* adapter) {
+  bool is_hardware = false;
+  THROW_IF_FAILED(adapter->GetProperty(
+    DXCoreAdapterProperty::IsHardware,
+    &is_hardware));
+  return is_hardware;
+}
+
+static bool IsGPU(IDXCoreAdapter* compute_adapter) {
+  // Only considering hardware adapters
+  if (!IsHardwareAdapter(compute_adapter)) {
+    return false;
+  }
+  return compute_adapter->IsAttributeSupported(DXCORE_ADAPTER_ATTRIBUTE_D3D12_GRAPHICS);
+}
+
+static bool IsNPU(IDXCoreAdapter* compute_adapter) {
+  // Only considering hardware adapters
+  if (!IsHardwareAdapter(compute_adapter)) {
+    return false;
+  }
+  return !(compute_adapter->IsAttributeSupported(DXCORE_ADAPTER_ATTRIBUTE_D3D12_GRAPHICS));
+}
+
+enum class DeviceType { GPU, NPU, BadDevice };
+
+static DeviceType FilterAdapterTypeQuery(IDXCoreAdapter* adapter, OrtDmlDeviceFilter filter) {
+  auto allow_gpus = (filter & OrtDmlDeviceFilter::Gpu) == OrtDmlDeviceFilter::Gpu;
+  if (IsGPU(adapter) && allow_gpus) {
+    return DeviceType::GPU;
+  }
+
+  auto allow_npus = (filter & OrtDmlDeviceFilter::Npu) == OrtDmlDeviceFilter::Npu;
+  if (IsNPU(adapter) && allow_npus) {
+    return DeviceType::NPU;
+  }
+
+  return DeviceType::BadDevice;
+}
+
+// Struct for holding each adapter
+struct AdapterInfo {
+  ComPtr<IDXCoreAdapter> Adapter;
+  DeviceType Type; // GPU or NPU
+};
+
+static ComPtr<IDXCoreAdapterList> EnumerateDXCoreAdapters(IDXCoreAdapterFactory* adapter_factory) {
+  ComPtr<IDXCoreAdapterList> adapter_list;
+
+  // use_dxcore_workload_enumeration should be determined by QI
```

Comment:
This is the spot that changes when DXCore work is done, right? Can you add a comment about that? (that we're waiting for work)